infrastructure prep for NAT reachability support

arifogel opened the pull request:
- break relations into basic and transformation-type
- break apart NodProgram for better visibility and abstraction
  - ReachabilityProgram - reachability program using only Batfish AST nodes
  - NodContext - Z3 declarations and context
  - NodProgram - Z3 rules, queries, and NodContext
- bunch of tests
- builders

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/batfish/batfish/960)
<!-- Reviewable:end -->

anothermattbrown commented:
Reviewed 91 of 91 files at r1.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[projects/batfish/src/main/java/org/batfish/z3/EarliestMoreGeneralReachableLineQuerySynthesizer.java, line 41 at r1](https://reviewable.io:443/reviews/batfish/batfish/960#-L5oH8OtdhVXyZX6IynA:-L5oH8OtdhVXyZX6IynB:bc7l7ny) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/EarliestMoreGeneralReachableLineQuerySynthesizer.java#L41)):*
> ```Java
> 
>   @Override
>   public ReachabilityProgram getReachabilityProgram(SynthesizerInput input) {
> ```

Add javadoc comment?

---

*[projects/batfish/src/main/java/org/batfish/z3/TransformationHeaderField.java, line 3 at r1](https://reviewable.io:443/reviews/batfish/batfish/960#-L5oc-W6pqCEOs6EWg-J:-L5oc-W6pqCEOs6EWg-K:bxv6xy5) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/TransformationHeaderField.java#L3)):*
> ```Java
> package org.batfish.z3;
> 
> public enum TransformationHeaderField implements HeaderField {
> ```

What is the need to have two enums of header field? Why not just keep them all in one?

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/DelegateBooleanExpr.java, line 11 at r1](https://reviewable.io:443/reviews/batfish/batfish/960#-L5oJbzCpTZYN6JNwcFr:-L5oJbzDoHV8EVYug-K5:b2s2ria) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/expr/DelegateBooleanExpr.java#L11)):*
> ```Java
> import org.batfish.z3.expr.visitors.ExprVisitor;
> 
> public abstract class DelegateBooleanExpr extends BooleanExpr {
> ```

What is this?

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/DelegateIntExpr.java, line 8 at r1](https://reviewable.io:443/reviews/batfish/batfish/960#-L5odmaGsob39OvrCkXi:-L5odmaGsob39OvrCkXj:bruj3ir) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/expr/DelegateIntExpr.java#L8)):*
> ```Java
> import org.batfish.z3.expr.visitors.IntExprVisitor;
> 
> public abstract class DelegateIntExpr extends IntExpr {
> ```

What is this? (add comment)

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/StateExpr.java, line 9 at r1](https://reviewable.io:443/reviews/batfish/batfish/960#-L5okrgzw80Jz5yiJicz:-L5okrgzw80Jz5yiJid-:b-gcoym9) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/expr/StateExpr.java#L9)):*
> ```Java
> import org.batfish.z3.state.visitors.StateVisitor;
> 
> public abstract class StateExpr extends BooleanExpr {
> ```

Why do we need this if it's not visitable?

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/TransformationRuleStatement.java, line 3 at r1](https://reviewable.io:443/reviews/batfish/batfish/960#-L5oiZ9BUrxE6kulQ_zk:-L5oiZ9BUrxE6kulQ_zl:b2s2ria) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/expr/TransformationRuleStatement.java#L3)):*
> ```Java
> package org.batfish.z3.expr;
> 
> public class TransformationRuleStatement extends RuleStatement {
> ```

What is this?

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/TransformationStateExpr.java, line 10 at r1](https://reviewable.io:443/reviews/batfish/batfish/960#-L5ohVnLO_jrNyAwYb_e:-L5ohVnLO_jrNyAwYb_f:b-epjinc) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/expr/TransformationStateExpr.java#L10)):*
> ```Java
> import org.batfish.z3.state.visitors.StateVisitor;
> 
> public abstract class TransformationStateExpr extends StateExpr {
> ```

What purpose does this serve? Why not let PostOutInterface extend StateExpr directly?

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/visitors/BoolExprTransformer.java, line 60 at r1](https://reviewable.io:443/reviews/batfish/batfish/960#-L5ojTFRePtb4Vfophkd:-L5ojTFRePtb4Vfophke:b-lsj3to) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/expr/visitors/BoolExprTransformer.java#L60)):*
> ```Java
>   private final Supplier<Expr[]> _basicArguments;
> 
>   /** Visible for delegate */
> ```

Why not make this the return value of the visit methods?

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/visitors/IsComplexVisitor.java, line 121 at r1](https://reviewable.io:443/reviews/batfish/batfish/960#-L5okP3aPIx484DttQzH:-L5okP3aPIx484DttQzI:byajovd) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/expr/visitors/IsComplexVisitor.java#L121)):*
> ```Java
>   }
> 
>   public void visitStateExpr(StateExpr stateExpr) {
> ```

What's this for if the method has been removed from ExprVisitor?

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/visitors/RelationCollector.java, line 164 at r1](https://reviewable.io:443/reviews/batfish/batfish/960#-L5olc5sWBAkoUJxtpa4:-L5olc5sWBAkoUJxtpa5:bt6hq4i) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/expr/visitors/RelationCollector.java#L164)):*
> ```Java
>   }
> 
>   public void visitRuleStatement(RuleStatement ruleStatement) {
> ```

Why is this here if it's not part of  VoidStatementVisitor?

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/visitors/RelationCollector.java, line 173 at r1](https://reviewable.io:443/reviews/batfish/batfish/960#-L5olpd4t9EoLvPlZL1b:-L5olpd4t9EoLvPlZL1c:ba3dfna) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/expr/visitors/RelationCollector.java#L173)):*
> ```Java
>   }
> 
>   public void visitStateExpr(StateExpr stateExpr) {
> ```

Remove?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/960)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 75 of 96 files reviewed at latest revision, 11 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/EarliestMoreGeneralReachableLineQuerySynthesizer.java, line 41 at r1](https://beta.reviewable.io/reviews/batfish/batfish/960#-L5oH8OtdhVXyZX6IynA:-L5pNcdjEGQg8b-OUTWU:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/EarliestMoreGeneralReachableLineQuerySynthesizer.java#L41)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Add javadoc comment?
</blockquote></details>

Done.

---

*[projects/batfish/src/main/java/org/batfish/z3/TransformationHeaderField.java, line 3 at r1](https://beta.reviewable.io/reviews/batfish/batfish/960#-L5oc-W6pqCEOs6EWg-J:-L5pHP08D-RNFOn-0dFZ:b9b7a35) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/TransformationHeaderField.java#L3)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

What is the need to have two enums of header field? Why not just keep them all in one?
</blockquote></details>

`TransformationHeaderField` has other members that `BasicHeaderField` does not.
Namely, a `TransformationHeaderField` points to the original and current header field of its type.
E.g. `NEW_DST_IP` points to `ORIG_DST_IP` and `DST_IP` respectively.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/DelegateBooleanExpr.java, line 11 at r1](https://beta.reviewable.io/reviews/batfish/batfish/960#-L5oJbzCpTZYN6JNwcFr:-L5pHxgGAcdzYVHWxOwT:bvxtuk) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/expr/DelegateBooleanExpr.java#L11)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

What is this?
</blockquote></details>

I wanted a way to make one or more simple `BooleanExpr` subclasses for testing that did not need to be known a priori by the relevant visitors. 
See `TestBooleanExpr`.
I added a comment to that effect.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/DelegateIntExpr.java, line 8 at r1](https://beta.reviewable.io/reviews/batfish/batfish/960#-L5odmaGsob39OvrCkXi:-L5pILbDD6icr2-wyIvH:b-q9tcw4) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/expr/DelegateIntExpr.java#L8)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

What is this? (add comment)
</blockquote></details>

I wanted a way to make one or more simple `IntExpr` subclasses for testing that did not need to be known a priori by the relevant visitors. 
See `TestIntExpr`.
I added a comment to that effect.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/StateExpr.java, line 9 at r1](https://beta.reviewable.io/reviews/batfish/batfish/960#-L5okrgzw80Jz5yiJicz:-L5pIZ4vAsrTOT-xV-rW:b8a2xsk) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/expr/StateExpr.java#L9)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Why do we need this if it's not visitable?
</blockquote></details>

What do you mean? There is a visitor right below your question.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/TransformationRuleStatement.java, line 3 at r1](https://beta.reviewable.io/reviews/batfish/batfish/960#-L5oiZ9BUrxE6kulQ_zk:-L5pIiV3AxkSSMsDbiZo:b-9h7686) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/expr/TransformationRuleStatement.java#L3)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

What is this?
</blockquote></details>

I broke RuleStatement into `BasicRuleStatement` and `TransformationRuleStatement`, based on whether the consequent is a `BasicStateExpr` or `TransformationStateExpr`.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/TransformationStateExpr.java, line 10 at r1](https://beta.reviewable.io/reviews/batfish/batfish/960#-L5ohVnLO_jrNyAwYb_e:-L5pJ7fv4IQICWD16jhN:btagsif) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/expr/TransformationStateExpr.java#L10)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

What purpose does this serve? Why not let PostOutInterface extend StateExpr directly?
</blockquote></details>

(via code added since your question)
Since `PostOutInterface` it extends `TransformationStateExpr`, the return value of its `getState()` must be a subclass of `TransformationState`.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/visitors/BoolExprTransformer.java, line 60 at r1](https://beta.reviewable.io/reviews/batfish/batfish/960#-L5ojTFRePtb4Vfophkd:-L5pQRrOEo71JlwjcyuC:b-sqkumc) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/expr/visitors/BoolExprTransformer.java#L60)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Why not make this the return value of the visit methods?
</blockquote></details>

That comment was actually wrong.
I removed it.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/visitors/IsComplexVisitor.java, line 121 at r1](https://beta.reviewable.io/reviews/batfish/batfish/960#-L5okP3aPIx484DttQzH:-L5pQZpH75TsVYzWLonz:bpfml6a) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/expr/visitors/IsComplexVisitor.java#L121)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

What's this for if the method has been removed from ExprVisitor?
</blockquote></details>

Code reuse. I made it `private` since it is only used internally.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/visitors/RelationCollector.java, line 164 at r1](https://beta.reviewable.io/reviews/batfish/batfish/960#-L5olc5sWBAkoUJxtpa4:-L5pQYzI0Tg-AIRaVsBN:bpfml6a) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/expr/visitors/RelationCollector.java#L164)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Why is this here if it's not part of  VoidStatementVisitor?
</blockquote></details>

Code reuse. I made it `private` since it is only used internally.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/visitors/RelationCollector.java, line 173 at r1](https://beta.reviewable.io/reviews/batfish/batfish/960#-L5olpd4t9EoLvPlZL1b:-L5pQzI4A6_Z86sT97Lw:bdaxl1w) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/expr/visitors/RelationCollector.java#L173)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Remove?
</blockquote></details>

It's here for code reuse. I made it `private` since it is only used internally.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/960)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Reviewed 19 of 29 files at r2.
Review status: 87 of 97 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/batfish/batfish/960#-L5tkRFlhL_SwJyqsoNU:-L5tkRFlhL_SwJyqsoNV:b-vx2v2f) (no related file):*
Add simplifier test for transformed expr

---

*[projects/batfish/src/main/java/org/batfish/z3/EarliestMoreGeneralReachableLineQuerySynthesizer.java, line 17 at r2](https://reviewable.io:443/reviews/batfish/batfish/960#-L5tgeZ8Wc0oXbgVJrb9:-L5tgeZ8Wc0oXbgVJrbA:brarzdx) ([raw file](https://github.com/batfish/batfish/blob/7ce94954bce206bc421c94f5500802108b159a3e/projects/batfish/src/main/java/org/batfish/z3/EarliestMoreGeneralReachableLineQuerySynthesizer.java#L17)):*
> ```Java
> import org.batfish.z3.state.NumberedQuery;
> 
> public class EarliestMoreGeneralReachableLineQuerySynthesizer
> ```

Comment?

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/visitors/BoolExprTransformer.java, line 236 at r2](https://reviewable.io:443/reviews/batfish/batfish/960#-L5tizRxiTBfH8KxjOR9:-L5tizRxiTBfH8KxjORA:b66uzcw) ([raw file](https://github.com/batfish/batfish/blob/7ce94954bce206bc421c94f5500802108b159a3e/projects/batfish/src/main/java/org/batfish/z3/expr/visitors/BoolExprTransformer.java#L236)):*
> ```Java
>     /* TODO: allow vectorized variables */
>     BoolExpr subExpression = toBoolExpr(transformedExpr.getSubExpression(), _input, _nodContext);
>     _boolExpr = (BoolExpr) subExpression.substitute(_from.get(), _to.get());
> ```

Add comment with simple example of a use.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/visitors/Simplifier.java, line 357 at r2](https://reviewable.io:443/reviews/batfish/batfish/960#-L5tjfUw1bYnr6xnrIBC:-L5tjfUw1bYnr6xnrIBD:bdnpc7x) ([raw file](https://github.com/batfish/batfish/blob/7ce94954bce206bc421c94f5500802108b159a3e/projects/batfish/src/main/java/org/batfish/z3/expr/visitors/Simplifier.java#L357)):*
> ```Java
>     BooleanExpr originalSubExpression = transformedExpr.getSubExpression();
>     BooleanExpr simplifiedSubExpression = simplifyBooleanExpr(originalSubExpression);
>     if (simplifiedSubExpression instanceof TransformedExpr) {
> ```

Transformation is idempotent. Add comment.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/visitors/Simplifier.java, line 361 at r2](https://reviewable.io:443/reviews/batfish/batfish/960#-L5tk1o_E-MsOGjtmiln:-L5tk1o_E-MsOGjtmilo:b-c59hrr) ([raw file](https://github.com/batfish/batfish/blob/7ce94954bce206bc421c94f5500802108b159a3e/projects/batfish/src/main/java/org/batfish/z3/expr/visitors/Simplifier.java#L361)):*
> ```Java
>     } else if (simplifiedSubExpression != originalSubExpression) {
>       _simplifiedBooleanExpr = new TransformedExpr(simplifiedSubExpression);
> ```

extra line

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/960)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Reviewed 10 of 29 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/DelegateBooleanExpr.java, line 11 at r1](https://reviewable.io:443/reviews/batfish/batfish/960#-L5oJbzCpTZYN6JNwcFr:-L5tm3554SfJRbF4C_Sx:bqu2q87) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/expr/DelegateBooleanExpr.java#L11)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

I wanted a way to make one or more simple `BooleanExpr` subclasses for testing that did not need to be known a priori by the relevant visitors. 
See `TestBooleanExpr`.
I added a comment to that effect.
</blockquote></details>

add a link to an example unit test where this is used?
also might be better to move this to the test folder and remove it from ExprVisitor. can add a new interface DelegateBooleanExprVisitor with only one visit method, and this can cast to that in its accept method.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/DelegateIntExpr.java, line 8 at r1](https://reviewable.io:443/reviews/batfish/batfish/960#-L5odmaGsob39OvrCkXi:-L5tmwya2zg_E03PKTrq:b-ia41wb) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/expr/DelegateIntExpr.java#L8)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

I wanted a way to make one or more simple `IntExpr` subclasses for testing that did not need to be known a priori by the relevant visitors. 
See `TestIntExpr`.
I added a comment to that effect.
</blockquote></details>

same comment as above

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/StateExpr.java, line 9 at r1](https://reviewable.io:443/reviews/batfish/batfish/960#-L5okrgzw80Jz5yiJicz:-L5tn6nIMUDxAhqfOion:bh4gmsk) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/expr/StateExpr.java#L9)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

What do you mean? There is a visitor right below your question.
</blockquote></details>

I meant there isn't a visitStateExpr method in BooleanExprVisitor. Maybe just add a comment why that is. What happens when BooleanExprVisitor tries to visit a StateExpr? Looks like either visitBasicStateExpr or visitTransformationStateExpr will be called. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/960)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[a discussion](https://beta.reviewable.io/reviews/batfish/batfish/960#-L5tkRFlhL_SwJyqsoNU:-L5tpYK2AowM5QstmUe2:b-896fix) (no related file):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Add simplifier test for transformed expr
</blockquote></details>

Done.

---

*[projects/batfish/src/main/java/org/batfish/z3/EarliestMoreGeneralReachableLineQuerySynthesizer.java, line 17 at r2](https://beta.reviewable.io/reviews/batfish/batfish/960#-L5tgeZ8Wc0oXbgVJrb9:-L5tqZ1nDJA-Vkw-BQMN:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/7ce94954bce206bc421c94f5500802108b159a3e/projects/batfish/src/main/java/org/batfish/z3/EarliestMoreGeneralReachableLineQuerySynthesizer.java#L17)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Comment?
</blockquote></details>

Done.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/DelegateBooleanExpr.java, line 11 at r1](https://beta.reviewable.io/reviews/batfish/batfish/960#-L5oJbzCpTZYN6JNwcFr:-L5trD0R02TE_8aovdnQ:b-l1brok) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/expr/DelegateBooleanExpr.java#L11)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

add a link to an example unit test where this is used?
also might be better to move this to the test folder and remove it from ExprVisitor. can add a new interface DelegateBooleanExprVisitor with only one visit method, and this can cast to that in its accept method.
</blockquote></details>

I cannot link to a class outside of the classpath. The whole point of this class is to provide a delegation mechanism for outside classes.

How would your suggestion work with e.g. `Simplifier`? That is, how could I make `Simplifier` delegate the simplification code for `TestBooleanExpr` outside of `Simplifier`?

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/StateExpr.java, line 9 at r1](https://beta.reviewable.io/reviews/batfish/batfish/960#-L5okrgzw80Jz5yiJicz:-L5tt8qa07HDjie3Yvc-:bd4wtvv) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/expr/StateExpr.java#L9)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

I meant there isn't a visitStateExpr method in BooleanExprVisitor. Maybe just add a comment why that is. What happens when BooleanExprVisitor tries to visit a StateExpr? Looks like either visitBasicStateExpr or visitTransformationStateExpr will be called. 
</blockquote></details>

I added a comment but it does not address visitation. Shouldn't the default for an abstract expression class be to not have a dedicated visit method in its visitors? I feel like a comment would only be warranted if it *had* one.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/visitors/BoolExprTransformer.java, line 236 at r2](https://beta.reviewable.io/reviews/batfish/batfish/960#-L5tizRxiTBfH8KxjOR9:-L5twaG6CyDzPcSDj0vO:b-ug0dx4) ([raw file](https://github.com/batfish/batfish/blob/7ce94954bce206bc421c94f5500802108b159a3e/projects/batfish/src/main/java/org/batfish/z3/expr/visitors/BoolExprTransformer.java#L236)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Add comment with simple example of a use.
</blockquote></details>

I added such a comment, but for now the citation relies on code yet to be written.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/visitors/Simplifier.java, line 357 at r2](https://beta.reviewable.io/reviews/batfish/batfish/960#-L5tjfUw1bYnr6xnrIBC:-L5tuERwA0xviOdL-2Yk:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/7ce94954bce206bc421c94f5500802108b159a3e/projects/batfish/src/main/java/org/batfish/z3/expr/visitors/Simplifier.java#L357)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Transformation is idempotent. Add comment.
</blockquote></details>

Done.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/visitors/Simplifier.java, line 361 at r2](https://beta.reviewable.io/reviews/batfish/batfish/960#-L5tk1o_E-MsOGjtmiln:-L5twrse8o06Xn_gCIbF:bss4a2b) ([raw file](https://github.com/batfish/batfish/blob/7ce94954bce206bc421c94f5500802108b159a3e/projects/batfish/src/main/java/org/batfish/z3/expr/visitors/Simplifier.java#L361)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

extra line
</blockquote></details>

Removed.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/960)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Reviewed 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/DelegateBooleanExpr.java, line 11 at r1](https://beta.reviewable.io/reviews/batfish/batfish/960#-L5oJbzCpTZYN6JNwcFr:-L5tzdgT7gXojTRu74ON:bqnjckq) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/expr/DelegateBooleanExpr.java#L11)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

I cannot link to a class outside of the classpath. The whole point of this class is to provide a delegation mechanism for outside classes.

How would your suggestion work with e.g. `Simplifier`? That is, how could I make `Simplifier` delegate the simplification code for `TestBooleanExpr` outside of `Simplifier`?
</blockquote></details>

`Simplifier` shouldn't have to delegate -- it can be delegated to. Delegation originates from `TestBooleanExpr.accept`. It will delegate to `DelegateBooleanExprVisitor`, which presumably has the `visit` method for `TestBooleanExpr`. It can either extend or wrap `Simplifier` for the other cases.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/StateExpr.java, line 9 at r1](https://beta.reviewable.io/reviews/batfish/batfish/960#-L5okrgzw80Jz5yiJicz:-L5u-yeC64i__MPC4rCO:b-f5li1p) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/expr/StateExpr.java#L9)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

I added a comment but it does not address visitation. Shouldn't the default for an abstract expression class be to not have a dedicated visit method in its visitors? I feel like a comment would only be warranted if it *had* one.
</blockquote></details>

makes sense.  Maybe add comments to `BasicStateExpr` and `TransformationStateExpr` then.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/960)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Reviewed 13 of 13 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[projects/batfish/src/main/java/org/batfish/z3/SynthesizerInput.java, line 67 at r4](https://reviewable.io:443/reviews/batfish/batfish/960#-L5uhbYZVaBFC1KQk6sY:-L5uhbYZVaBFC1KQk6sZ:bep8zt9) ([raw file](https://github.com/batfish/batfish/blob/dfeb195a30b6bebe48b06ff30a55fb34d04959df/projects/batfish/src/main/java/org/batfish/z3/SynthesizerInput.java#L67)):*
> ```Java
> 
>   /** Mapping: hostname -> interface -> [(matchCondition, transformationToApply)] */
>   Map<String, Map<String, List<Entry<BooleanExpr, BooleanExpr>>>> getSourceNats();
> ```

Avoid using Entry -- adding a class will make the code more readable. 

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 807 at r4](https://reviewable.io:443/reviews/batfish/batfish/960#-L5uinyitBW8AO2Vihfg:-L5uinyitBW8AO2Vihfh:b-bggl9x) ([raw file](https://github.com/batfish/batfish/blob/dfeb195a30b6bebe48b06ff30a55fb34d04959df/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L807)):*
> ```Java
>                               ImmutableList.<BooleanExpr>builder()
>                                   .add(sourceNats.get(i).getKey())
>                                   .add(sourceNats.get(i).getValue());
> ```

Here's where a better type would make this more readable.
getKey gives us the match condition of the NAT
getValue gives us the update rule. Should this be wrapped in a transformation expr?

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 814 at r4](https://reviewable.io:443/reviews/batfish/batfish/960#-L5ujSlBv1L8wBJRU1BQ:-L5ujSlBv1L8wBJRU1BR:b-665ctw) ([raw file](https://github.com/batfish/batfish/blob/dfeb195a30b6bebe48b06ff30a55fb34d04959df/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L814)):*
> ```Java
>                                 .map(Entry::getKey)
>                                 .map(NotExpr::new)
>                                 .forEach(currentSourceNatConjuncts::add);
> ```

Rather than retraversing the list each iteration, can accumulate the negations of previous match conditions in a single `BooleanExpr`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/960)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 82 of 118 files reviewed at latest revision, 5 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/DelegateBooleanExpr.java, line 11 at r1](https://beta.reviewable.io/reviews/batfish/batfish/960#-L5oJbzCpTZYN6JNwcFr:-L5utxJw65TnWZUs8dbl:b-pzqnxj) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/expr/DelegateBooleanExpr.java#L11)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

`Simplifier` shouldn't have to delegate -- it can be delegated to. Delegation originates from `TestBooleanExpr.accept`. It will delegate to `DelegateBooleanExprVisitor`, which presumably has the `visit` method for `TestBooleanExpr`. It can either extend or wrap `Simplifier` for the other cases.
</blockquote></details>

I refactored the visitors in a way that allowed me to eliminate delegate classes.

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/StateExpr.java, line 9 at r1](https://beta.reviewable.io/reviews/batfish/batfish/960#-L5okrgzw80Jz5yiJicz:-L5uvcAhCAO3i34ILkrK:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/d3911b1b0ab52dc6f934ddcee39bd219161770df/projects/batfish/src/main/java/org/batfish/z3/expr/StateExpr.java#L9)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

makes sense.  Maybe add comments to `BasicStateExpr` and `TransformationStateExpr` then.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/960)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 82 of 118 files reviewed at latest revision, 4 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/SynthesizerInput.java, line 67 at r4](https://beta.reviewable.io/reviews/batfish/batfish/960#-L5uhbYZVaBFC1KQk6sY:-L5uw_lJ5ondu3a14M2N:b-aghy61) ([raw file](https://github.com/batfish/batfish/blob/dfeb195a30b6bebe48b06ff30a55fb34d04959df/projects/batfish/src/main/java/org/batfish/z3/SynthesizerInput.java#L67)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Avoid using Entry -- adding a class will make the code more readable. 
</blockquote></details>

I disagree that using a class other than `Entry` will make this any more readable given the already heavily-nested generic return type.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 807 at r4](https://beta.reviewable.io/reviews/batfish/batfish/960#-L5uinyitBW8AO2Vihfg:-L5uxOVKBEziBSM3ZHrW:bjmkc5a) ([raw file](https://github.com/batfish/batfish/blob/dfeb195a30b6bebe48b06ff30a55fb34d04959df/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L807)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Here's where a better type would make this more readable.
getKey gives us the match condition of the NAT
getValue gives us the update rule. Should this be wrapped in a transformation expr?
</blockquote></details>

I compromised by using descriptive variables.

---

*[projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java, line 814 at r4](https://beta.reviewable.io/reviews/batfish/batfish/960#-L5ujSlBv1L8wBJRU1BQ:-L5uyBZ8BWCXe67_24Ky:bllbztt) ([raw file](https://github.com/batfish/batfish/blob/dfeb195a30b6bebe48b06ff30a55fb34d04959df/projects/batfish/src/main/java/org/batfish/z3/state/visitors/DefaultTransitionGenerator.java#L814)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Rather than retraversing the list each iteration, can accumulate the negations of previous match conditions in a single `BooleanExpr`
</blockquote></details>

In practice I have never seen more than 2 source NAT entries for an interface, so the performance gain would likely be negligible. Furthermore, this is not exactly an expensive operation per entry.  Ans since I believe your suggestion will make the code harder to read, I would prefer to leave as is.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/960)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Reviewed 38 of 39 files at r5.
Review status: 117 of 118 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/960)*
<!-- Sent from Reviewable.io -->

anothermattbrown commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/960)*
<!-- Sent from Reviewable.io -->
